delocate/tools.py

```diff
@@ -163,9 +173,136 @@ def parse_install_name(line):
         compatibility version
     current_version : str
         current version
-    """
+
+    Examples
+    --------
+    >>> parse_install_name("/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 905.6.0)")
+    ('/usr/lib/libc++.1.dylib', '1.0.0', '905.6.0')
+    """  # noqa: E501
     line = line.strip()
-    return IN_RE.match(line).groups()
+    match = IN_RE.match(line)
+    if not match:
+        raise ValueError(f"Could not parse {line!r}")
+    libname, compat_version, current_version = match.groups()
+    return libname, compat_version, current_version
+
+
+_OTOOL_ARCHITECTURE_RE = re.compile(
+    r"^(?P<name>.*?)(?: \(architecture (?P<architecture>\w+)\))?:$"
+)
+"""Matches the library separater line in 'otool -L'-like outputs.
+
+Examples
+--------
+>>> match = _OTOOL_ARCHITECTURE_RE.match("example.so (architecture x86_64):")
+>>> match["name"]
+'example.so'
+>>> match["architecture"]
+'x86_64'
+>>> match = _OTOOL_ARCHITECTURE_RE.match("example.so:")
+>>> match["name"]
+'example.so'
+>>> match["architecture"] is None
+True
+"""
+
+
+def _parse_otool_listing(stdout: str) -> Dict[str, List[str]]:
+    '''Parse the output of otool lists.
+
+    Parameters
+    ----------
+    stdout : str
+        A decoded stdout of commands like 'otool -L' or 'otool -D' to be parsed.
+
+    Returns
+    -------
+    otool_out : dict of list of str
+        The dictionary key is the architecture, e.g. 'x86_64'.
+        If no architecture is parsed then the key is ''.
+        The values are a list of strings associated with the key.
+
+    Examples
+    --------
+    >>> _parse_otool_listing("""
+    ... example.so (architecture x86_64):
+    ... \titem_1
+    ... \titem_2
+    ... example.so (architecture arm64):
+    ... \titem_3
+    ... \titem_4
+    ... """)
+    {'x86_64': ['item_1', 'item_2'], 'arm64': ['item_3', 'item_4']}
+    >>> _parse_otool_listing("""
+    ... example.so:
+    ... \titem_1
+    ... """)
+    {'': ['item_1']}
+    >>> _parse_otool_listing("""
+    ... example.so (architecture x86_64):
+    ... example.so (architecture arm64):
+    ... """)
+    {'x86_64': [], 'arm64': []}
+    '''
+    stdout = stdout.strip()
+    out: Dict[str, List[str]] = {}
+    lines = stdout.split("\n")
+    while lines:
+        # Detect and parse the name/arch header line.
+        match_arch = _OTOOL_ARCHITECTURE_RE.match(lines.pop(0))
+        assert match_arch, f"Missing file/architecture header:\n{stdout}"
```

Comment:
Asserts to errors (asserts will be optimized away by Python -0 flags)?

Reply:
A lot of these asserts are sanity checks.  I think I can turn most of these into `RuntimeError`.